llama_stack/providers/adapters/datasetio/huggingface/huggingface.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+from typing import List, Optional
+
+from llama_stack.apis.datasetio import *  # noqa: F403
+
+
+import datasets as hf_datasets
+from llama_stack.providers.datatypes import DatasetsProtocolPrivate
+from llama_stack.providers.utils.datasetio.url_utils import get_dataframe_from_url
+
+from .config import HuggingfaceDatasetIOConfig
+
+
+def load_hf_dataset(dataset_def: DatasetDef):
+    if dataset_def.metadata.get("path", None):
```

Comment:
it is a little unclear why this fork exists and is not clear to the user that putting `path` in the `metadata` will execute this codepath. i might be pre-coffee-braindead also. why don't we _always_ use `hf_datasets` or the other way?

Reply:
There are 2 branches in `load_hf_dataset` b/c huggingface provider allows (1) getting dataset from huggingface hub (2) user uploaded dataset. 

1. User have a dataset hosted on huggingface hub. User can access this dataset via specifying the `metadata` for getting huggingface dataset hub. For example: loading from https://huggingface.co/datasets/llamastack/evals

```python
mmlu = DatasetDefWithProvider(
            identifier="mmlu",
            url=URL(uri="https://huggingface.co/datasets/llamastack/evals"),
            dataset_schema={
                "input_query": StringType(),
                "expected_answer": StringType(),
                "chat_completion_input": ChatCompletionInputType(),
            },
            metadata={
                "path": "llamastack/evals",
                "name": "evals__mmlu__details",
                "split": "train",
            },
            provider_id="remote::huggingface",
        )
```

3. User have a dataset uploaded via URL. This is useful for registering custom dataset. This is the second path using `dataset = hf_datasets.Dataset.from_pandas(df)`. We can then additionally upload this dataset via to the remote huggingface repo. I didn't include the upload arguments in this PR, but they can be either stored in DatasetDef's metadata like the upload path or in a separate huggingface data structure.